src/app/reporting/Read-Checked.cpp

```diff
@@ -0,0 +1,140 @@
+            EmberImpl::RetrieveClusterData(dataModel, subjectDescriptor, isFabricFiltered, reportBuilder, path, encoderState);
+        lengthWrittenEmber = reportBuilder.GetWriter()->GetLengthWritten();
+    }
+
+    CHIP_ERROR errDM =
+        DataModelImpl::RetrieveClusterData(dataModel, subjectDescriptor, isFabricFiltered, reportBuilder, path, &stateDm);
+
+    if (errEmber != errDM)
+    {
+        // Note log + chipDie instead of VerifyOrDie so that breakpoints (and usage of rr)
+        // is easier to debug.
+        ChipLogError(Test, "Different return codes between ember and DM");
+        ChipLogError(Test, "  Ember error: %" CHIP_ERROR_FORMAT, errEmber.Format());
+        ChipLogError(Test, "  DM error: %" CHIP_ERROR_FORMAT, errDM.Format());
+        chipDie();
+    }
+
+    // data should be identical for most cases EXCEPT that for time-deltas (e.g. seconds since boot or similar)
+    // it may actually differ. As a result, the amount of data written in bytes MUST be the same, however if the rest of the
+    // data is not the same, we just print it out as a warning for manual inspection
+    //
+    // We have no direct access to TLV buffer data (especially given backing store splits)
+    // so for now we check that data length was identical.
+    //
+    // NOTE: RetrieveClusterData is responsible for encoding StatusIB errors in case of failures
+    //       so we validate length written requirements for BOTH success and failure.
+    //
+    // NOTE: data length is NOT reliable if the data content differs in encoding lentgh. E.g. numbers changing
+    //       from 0xFF to 0x100 or similar will use up more space.
+    //       For unit tests we make the validation strict, however for runtime we just report an
+    //       error for different sizes.
+    if (lengthWrittenEmber != reportBuilder.GetWriter()->GetLengthWritten())
+    {
+        ChipLogError(Test, "Different written length: %" PRIu32 " (Ember) vs %" PRIu32 " (DataModel)", lengthWrittenEmber,
+                     reportBuilder.GetWriter()->GetLengthWritten());
+#if CONFIG_BUILD_FOR_HOST_UNIT_TEST
+        chipDie();
+#endif
+    }
+
+    // For write resumes, the encoder state MUST be idential
+    if ((errEmber == CHIP_ERROR_NO_MEMORY) || (errEmber == CHIP_ERROR_BUFFER_TOO_SMALL))
+    {
+        // Encoder state MUST match on partial reads (used by chunking)
+        // specifically ReadViaAccessInterface in ember-compatibility-functions only
+        // sets the encoder state in case an error occurs.
+        if (encoderState != nullptr)
+        {
+            if (encoderState->AllowPartialData() != stateDm.AllowPartialData())
+            {
+                ChipLogError(Test, "Different partial data");
+                chipDie();
+            }
+            if (encoderState->CurrentEncodingListIndex() != stateDm.CurrentEncodingListIndex())
+            {
+                ChipLogError(Test, "Different partial data");
+                chipDie();
```

Comment:
This could get hit if you have a list of time-dependent values, so in one version they fit in the packet and in the other they do not....

Reply:
Moved to have all chipDie to unit tests only and added comments.